#### Imports

In [3]:
import pandas as pd
from random import randint
pd.set_option('display.max_rows', 5)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 10000)

#### Download the latest [Pleiades names csv file](https://atlantides.org/downloads/pleiades/dumps/) and read it.

In [20]:
df = pd.read_csv("pleiades-places-latest.csv")

In [21]:
df

authors                                            bbox connectsWith               created          creators  currentVersion                                        description                                             extent featureTypes geoContext hasConnectionsWith         id locationPrecision  maxDate  minDate              modified               path    reprLat            reprLatLong   reprLong                               tags timePeriods                                 timePeriodsKeys timePeriodsRange                   title                               uid
0      Becker, J., T. Elliott    13.4119837, 42.082885, 13.4119837, 42.082885       413005  2016-11-04T16:36:09Z  jbecker, thomase             1.0  The post-Roman settlement at Alba Fucens becam...  {"type": "Point", "coordinates": [13.4119837, ...   settlement        NaN                NaN   48210385           precise   1453.0    640.0  2016-11-08T21:58:28Z   /places/48210385  42.082885   42.082885,13.4119837  13.411984                                NaN           M                             mediaeval-byzantine     640.0,1453.0         Borgo Medievale  ece5760c4c6d42c1a331aad543c4ecc4
1      Becker, J., T. Elliott  11.6285463, 42.4193742, 11.6285463, 42.4193742       413393  2016-11-04T16:39:09Z  jbecker, thomase             2.0  A major urban sanctuary at Vulci with a long p...  {"type": "Point", "coordinates": [11.6285463, ...     temple-2        NaN                NaN   48210386           precise    300.0   -750.0  2016-12-05T11:47:10Z   /places/48210386  42.419374  42.4193742,11.6285463  11.628546  sanctuary, extant remains, temple        ACHR  archaic,classical,hellenistic-republican,roman     -750.0,300.0  Tempio Grande at Vulci  4e06898f2de74dbc9f3a3bdba6d74ba2
...                       ...                                             ...          ...                   ...               ...             ...                                                ...                                                ...          ...        ...                ...        ...               ...      ...      ...                   ...                ...        ...                    ...        ...                                ...         ...                                             ...              ...                     ...                               ...
40372  Becker, J., T. Elliott  -7.6859672, 42.9800949, -7.6859672, 42.9800949          NaN  2014-01-13T04:47:37Z           jbecker             3.0  The Roman sanctuary of Santalla de Bóveda date...  {"type": "Point", "coordinates": [-7.6859672, ...     temple-2        NaN                NaN  743134062           precise    640.0    -30.0  2016-06-30T16:47:31Z  /places/743134062  42.980095  42.9800949,-7.6859672  -7.685967                                NaN          RL                              roman,late-antique      -30.0,640.0      Santalla de Bóveda  a727adadcde6eb313a9c6af1564e973b
40373              Becker, J.  -8.1303178, 42.9308198, -8.1303178, 42.9308198          NaN  2015-05-23T04:33:55Z           jbecker             1.0  A Roman bridge located in the Provincia da Cor...  {"type": "Point", "coordinates": [-8.1303178, ...       bridge        NaN                NaN  623510547           precise    640.0    -30.0  2021-10-10T13:04:32Z  /places/623510547  42.930820  42.9308198,-8.1303178  -8.130318                                NaN           R                 roman-early-empire-late-antique      -30.0,640.0   Ribadiso Roman bridge  8fd3386ba838b5bf1ef88a0f1a7df896

[40374 rows x 26 columns]

#### Print the first two rows

In [19]:
df.head(2)

authors                                        bbox               created    creators  currentVersion description                                             extent         id locationPrecision  maxDate  minDate              modified nameAttested nameLanguage nameTransliterated                      path             pid    reprLat          reprLatLong  reprLong tags timePeriods                            timePeriodsKeys timePeriodsRange      title                               uid
0  Spann, P., R. Warner, R. Talbert, T. Elliott, ...  -3.606772, 39.460299, -3.606772, 39.460299  2010-09-24T19:02:22Z  P.O. Spann             NaN         NaN  {"type": "Point", "coordinates": [-3.606772, 3...  consabura           precise    640.0   -330.0  2011-09-05T20:57:22Z          NaN          NaN          Consabura  /places/265876/consabura  /places/265876  39.460299  39.460299,-3.606772 -3.606772  NaN         HRL  hellenistic-republican,roman,late-antique     -330.0,640.0  Consabura  fc0ee157ff11ce6d2e72cd7c5df06fee
1  Spann, P., R. Warner, R. Talbert, T. Elliott, ...  -3.606772, 39.460299, -3.606772, 39.460299  2010-09-24T19:02:22Z  P.O. Spann             NaN         NaN  {"type": "Point", "coordinates": [-3.606772, 3...  consabrum           precise    640.0   -330.0  2011-09-05T20:57:22Z          NaN          NaN          Consabrum  /places/265876/consabrum  /places/265876  39.460299  39.460299,-3.606772 -3.606772  NaN         HRL  hellenistic-republican,roman,late-antique     -330.0,640.0  Consabrum  e2b8756302fb62ff0e301710c265a3e6

#### Use the `describe()` method to show basic statistics (mean, median, std, etc.) of numeric columns

In [14]:
df.describe()

currentVersion       maxDate       minDate       reprLat      reprLong
count    19159.000000  33793.000000  33793.000000  18767.000000  18767.000000
mean         0.661465    668.623028  -1144.212648     38.149300     24.329923
...               ...           ...           ...           ...           ...
75%          1.000000    640.000000    300.000000     41.906964     35.231201
max         16.000000   2100.000000   2000.000000     62.500000    103.866580

[8 rows x 5 columns]

#### Print the name of the columns

In [15]:
df.columns

Index(['authors', 'bbox', 'created', 'creators', 'currentVersion', 'description', 'extent', 'id', 'locationPrecision', 'maxDate', 'minDate', 'modified', 'nameAttested', 'nameLanguage', 'nameTransliterated', 'path', 'pid', 'reprLat', 'reprLatLong', 'reprLong', 'tags', 'timePeriods', 'timePeriodsKeys', 'timePeriodsRange', 'title', 'uid'], dtype='object')

#### Find how many _uniques_ values are in the featureTypes column.

In [23]:
df["featureTypes"].nunique()

1235

Result: 1235

#### Find Athens
Hint: It is recorded as "Athenae."

In [31]:
 df[df["title"] == "Athenae"]

authors                                        bbox connectsWith               created     creators  currentVersion                                        description                                             extent featureTypes     geoContext                                 hasConnectionsWith      id locationPrecision  maxDate  minDate              modified            path    reprLat                reprLatLong   reprLong                                               tags timePeriods                                    timePeriodsKeys timePeriodsRange    title                               uid
22109  Traill, J., B. Kiesling, S. Gillies, C. Lee, J...  23.619913, 37.937222, 23.849968, 38.019497       579888  2018-06-07T20:56:36Z  J.S. Traill            28.0  A major Greek city-state and the principal cit...  {"type": "Polygon", "coordinates": [[[23.64460...   settlement  Athina/Athens  585128,728329644,235795850,580123,813654446,58...  579885           related   2100.0   -750.0  2022-08-25T22:21:35Z  /places/579885  37.972454  37.972453603,23.726268222  23.726268  dare:ancient=1, dare:major=1, dare:feature=maj...      ACHRLM  archaic,classical,hellenistic-republican,roman...    -750.0,2100.0  Athenae  aef8271177d5afba233b2b97dcb52af4

#### Find all sanctuaries.
Hint: Have a look first in column "featureTypes". There are more than one piece of information in some cells. How does this affect our code? We have to search for cells with strings that contain the string we are looking for.

In [107]:
df[df["featureTypes"].str.contains("sanctuary", na=False)]

authors                                            bbox connectsWith               created                                  creators  currentVersion                                        description                                             extent       featureTypes                      geoContext hasConnectionsWith         id locationPrecision  maxDate  minDate              modified               path    reprLat            reprLatLong   reprLong                                               tags timePeriods                                   timePeriodsKeys timePeriodsRange                     title                               uid
19     Spann, P., DARMC, R. Talbert, R. Warner, J. Be...      -1.143263, 38.208848, -1.143263, 38.208848          NaN  2010-09-24T19:02:53Z                                P.O. Spann             5.0      An ancient place, cited: BAtlas 27 D3 Fortuna  {"type": "Point", "coordinates": [-1.143263, 3...          sanctuary  Cueva Negra, 13 miles N Murcia                NaN     265899           precise    300.0   -330.0  2021-10-25T13:19:40Z     /places/265899  38.208848    38.208848,-1.143263  -1.143263  dare:ancient=1, dare:major=0, dare:feature=temple          HR                      hellenistic-republican,roman     -330.0,300.0                   Fortuna  7744c1ce1aef567d7e36bdf87400c205
1006   Fentress, E., DARMC, R. Talbert, R. Warner, J....        6.451106, 36.328033, 6.451106, 36.328033          NaN  2010-11-10T18:40:19Z                           E.W.B. Fentress             5.0     An ancient place, cited: BAtlas 31 E4 Chettaba  {"type": "Point", "coordinates": [6.451106, 36...    sanctuary, cave          Chettaba, sacred caves                NaN     305060           precise    300.0   -330.0  2021-11-15T15:45:23Z     /places/305060  36.328033     36.328033,6.451106   6.451106  dare:major=0, dare:ancient=0, dare:feature=temple          HR                      hellenistic-republican,roman     -330.0,300.0                  Chettaba  c359c9a2335230f8014143615fcbda12
...                                                  ...                                             ...          ...                   ...                                       ...             ...                                                ...                                                ...                ...                             ...                ...        ...               ...      ...      ...                   ...                ...        ...                    ...        ...                                                ...         ...                                               ...              ...                       ...                               ...
39866  Walter, D., E. Brod, Y. Yilmaz, J. Becker, T. ...    29.1270484, 37.926248, 29.1270484, 37.926248       638860  2018-03-15T17:19:12Z  Douglas Walter, Emily Brod, Yeşim Yilmaz            20.0  The Ploutonion at Hierapolis is a walled sanct...  {"type": "Point", "coordinates": [29.1270484, ...  sanctuary, shrine                             NaN                NaN  232079424           precise   2100.0   -330.0  2020-01-13T16:42:58Z  /places/232079424  37.926248   37.926248,29.1270484  29.127048                       sanctuary, Kore, cave, Pluto        HRLM  hellenistic-republican,roman,late-antique,modern    -330.0,2100.0  Ploutonion at Hierapolis  95af82bbb8994ffea012c2fa70d64452
39885                                         Becker, J.  -0.1894608, 52.5747722, -0.1894608, 52.5747722          NaN  2013-09-01T00:43:04Z                                   jbecker             2.0  Flag Fen is a Bronze Age ritual site near Pete...  {"type": "Point", "coordinates": [-0.1894608, ...          sanctuary                             NaN                NaN  535275912           precise   -800.0  -2500.0  2021-02-02T23:38:25Z  /places/535275912  52.574772  52.5747722,-0.1894608  -0.189461                                     extant remains           B                   

You may get an error 'ValueError: Cannot mask with non-boolean array containing NA / NaN values'. If yes, then use the argument `na=False` within the `str.contains()` function.

Result: 606 rows × 26 columns

#### Find all temples and sanctuaries
Hint: Have a look first in column "featureTypes". There are more than one piece of information in some cells. How does this affect your code?

In [108]:
df[df["featureTypes"].str.contains("sanctuary|temple", na=False)]

authors                                            bbox connectsWith               created          creators  currentVersion                                        description                                             extent featureTypes                      geoContext hasConnectionsWith         id locationPrecision  maxDate  minDate              modified               path    reprLat                 reprLatLong   reprLong                                               tags timePeriods                                    timePeriodsKeys timePeriodsRange                   title                               uid
1                                 Becker, J., T. Elliott  11.6285463, 42.4193742, 11.6285463, 42.4193742       413393  2016-11-04T16:39:09Z  jbecker, thomase             2.0  A major urban sanctuary at Vulci with a long p...  {"type": "Point", "coordinates": [11.6285463, ...     temple-2                             NaN                NaN   48210386           precise    300.0   -750.0  2016-12-05T11:47:10Z   /places/48210386  42.419374       42.4193742,11.6285463  11.628546                  sanctuary, extant remains, temple        ACHR     archaic,classical,hellenistic-republican,roman     -750.0,300.0  Tempio Grande at Vulci  4e06898f2de74dbc9f3a3bdba6d74ba2
19     Spann, P., DARMC, R. Talbert, R. Warner, J. Be...      -1.143263, 38.208848, -1.143263, 38.208848          NaN  2010-09-24T19:02:53Z        P.O. Spann             5.0      An ancient place, cited: BAtlas 27 D3 Fortuna  {"type": "Point", "coordinates": [-1.143263, 3...    sanctuary  Cueva Negra, 13 miles N Murcia                NaN     265899           precise    300.0   -330.0  2021-10-25T13:19:40Z     /places/265899  38.208848         38.208848,-1.143263  -1.143263  dare:ancient=1, dare:major=0, dare:feature=temple          HR                       hellenistic-republican,roman     -330.0,300.0                 Fortuna  7744c1ce1aef567d7e36bdf87400c205
...                                                  ...                                             ...          ...                   ...               ...             ...                                                ...                                                ...          ...                             ...                ...        ...               ...      ...      ...                   ...                ...        ...                         ...        ...                                                ...         ...                                                ...              ...                     ...                               ...
40352                              Mckee, G., T. Elliott  45.8867685, 31.6671599, 45.8880979, 31.6685474     44626252  2019-08-30T15:55:07Z            gmckee             2.0  Large temple at Tell Jokha/Umma dedicated to t...  {"type": "Polygon", "coordinates": [[[45.88718...     temple-2                             NaN                NaN  537242705           precise  -1600.0  -2335.0  2019-08-30T21:05:58Z  /places/537242705  31.667848  31.6678479795,45.887425241  45.887425                                                NaN          AO  akkadian-ur-iii-mesopotamia,old-babylonian-ass...  -2335.0,-1600.0    Shara Temple at Umma  0ea6ed47045c4f38a01c2a6626d14d24
40372                             Becker, J., T. Elliott  -7.6859672, 42.9800949, -7.6859672, 42.9800949          NaN  2014-01-13T04:47:37Z           jbecker             3.0  The Roman sanctuary of Santalla de Bóveda date...  {"type": "Point", "coordinates": [-7.6859672, ...     temple-2                             NaN                NaN  743134062           precise    640.0    -30.0  2016-06-30T16:47:31Z  /places/743134062  42.980095       42.9800949,-7.6859672  -7.685967                                                NaN          RL                                 roman,late-antique      -30.0,640.0      Santalla de Bóveda  a727adadcde6eb313a9c6af1564e973b

[1454 rows x 26 columns]

#### Find which is the northernest, southernest, easternest, and westernest sanctuary of the database.

##### northernest

In [84]:
df[df["reprLat"] == df["reprLat"].max()]

authors                   bbox connectsWith               created               creators  currentVersion                                        description                                             extent featureTypes geoContext hasConnectionsWith     id locationPrecision  maxDate  minDate              modified           path  reprLat reprLatLong  reprLong tags timePeriods timePeriodsKeys timePeriodsRange               title                               uid
29964  Warner, R., A. Bursche, R. Talbert, S. Gillies...  5.0, 60.0, 10.0, 65.0          NaN  2010-09-22T20:46:22Z  R. Warner, A. Bursche             3.0  The sixth century AD historian Iordanes descri...  {"type": "Polygon", "coordinates": [[[5.0, 60....       region    NOR/SWE                NaN  20595             rough    640.0    300.0  2013-10-27T15:41:29Z  /places/20595     62.5    62.5,7.5       7.5  NaN           L    late-antique      300.0,640.0  Scadinavia/Scandza  ea9d282bba80ccfc5c21e40562bc3d6b

##### southernest

In [85]:
df[df["reprLat"] == df["reprLat"].min()]

authors                                              bbox connectsWith               created                                         creators  currentVersion                                        description                                             extent                   featureTypes          geoContext hasConnectionsWith      id locationPrecision  maxDate  minDate              modified            path   reprLat                   reprLatLong   reprLong tags timePeriods                     timePeriodsKeys timePeriodsRange      title                               uid
1774  Hausleiter, A., M. Roaf, St J. Simpson, R. Wen...  31.1406026, -29.5443422, 31.1411004, -29.5432314          NaN  2022-07-26T20:32:13Z  A. Hausleiter, M. Roaf, St J. Simpson, R. Wenke             4.0  Kas(h)kar was a significant Sasanian city, ori...  {"type": "Polygon", "coordinates": [[[31.14060...  settlement, settlement-modern  opposite Wasit IRQ                NaN  912880           precise   2099.0    -30.0  2022-07-27T02:08:12Z  /places/912880 -29.54377  -29.5437696898,31.1408074997  31.140807  NaN         RLT  roman,late-antique,twenty-first-ce     -30.0,2099.0  Kas(h)kar  91cbfced9f7047504fe506c711ac3862

##### easternest

In [87]:
df[df["reprLong"] == df["reprLong"].max()]

authors                                             bbox connectsWith               created creators  currentVersion                                        description                                             extent featureTypes geoContext hasConnectionsWith         id locationPrecision  maxDate  minDate              modified               path   reprLat                 reprLatLong  reprLong            tags timePeriods                                    timePeriodsKeys timePeriodsRange         title                               uid
24655  Mckee, G., J. Becker  111.077259, 48.5376447, 111.0793082, 48.5392573          NaN  2017-11-01T18:00:55Z   gmckee             3.0  Xiongnu tomb site near the modern village of B...  {"type": "Polygon", "coordinates": [[[111.0784...         tomb        NaN                NaN  475881408           precise    250.0   -330.0  2020-11-07T03:26:08Z  /places/475881408  48.53836  48.5383602338,111.07830005  111.0783  UWHS Submitted        HKFS  hellenistic-central-asia,kangju-yuezhi-kushan-...     -330.0,250.0  Duurlig Nars  b533f3c95f904e12bc791209cd4650a1

##### westernest

In [88]:
df[df["reprLong"] == df["reprLong"].min()]

authors                                            bbox connectsWith               created             creators  currentVersion                                        description                                             extent                  featureTypes geoContext hasConnectionsWith      id locationPrecision  maxDate  minDate              modified            path    reprLat            reprLatLong    reprLong                                               tags timePeriods     timePeriodsKeys timePeriodsRange         title                               uid
2541  Pearce, M., P. Tozzi, DARMC, R. Talbert, S. Gi...  -117.221676, 34.132162, -117.221676, 34.132162          NaN  2010-11-15T17:54:49Z  M. Pearce, P. Tozzi             8.0  An ancient place, cited: BAtlas 39 D5 Alba Doc...  {"type": "Point", "coordinates": [-117.221676,...  station, archaeological-site   Albisola                NaN  383563           precise    640.0    -30.0  2023-04-04T23:42:38Z  /places/383563  34.132162  34.132162,-117.221676 -117.221676  dare:ancient=1, dare:major=0, dare:feature=sta...          RL  roman,late-antique      -30.0,640.0  Alba Docilia  ca9b667dcc6764585883ac16d764a08c

---

#### By using the methin `isin()`, check which items of list1 are _not_ in list2.
Hint: The following code creates two list of numbers from 1 to 100, removes randomly one number from each of them, and transforms the lists to Series.

In [90]:
list1 = [i for i in range(1,101)]
list1.pop(randint(1, 101))
ser1 = pd.Series(list1)
list2 = [i for i in range(1,101)]
list2.pop(randint(1, 101))
ser2 = pd.Series(list2)

In [110]:
#check by removing
list1 = [i for i in range(1, 101)]
list1.pop(randint(0, 99)) 
ser1 = pd.Series(list1)

list2 = [i for i in range(1, 101)]
list2.pop(randint(0, 99))  
ser2 = pd.Series(list2)

ser1[~ser1.isin(ser2)]

84    86
dtype: int64

---